File: IT/Multithreading/C++/farm.cpp
/* 
file farm.cpp 
compilare con: g++ -std=c++11 farm.cpp -lpthread
*/

#include <iostream>
#include <random>
#include <chrono>
#include <thread>
#include <condition_variable>

using std::cout;
using std::endl;
using std::ref;
using std::random_device;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::chrono::seconds;
using std::this_thread::sleep_for;
using std::thread;
using std::mutex;
using std::unique_lock;
using std::condition_variable;

void heap(int& bales, mutex& _mutex, condition_variable& condition, const char* tname)
{
	static thread_local default_random_engine gen(random_device{}());

	for (int i = 0; i < 3; i++) {
		unique_lock<mutex> lock(_mutex);

		while (bales > 0)
		{
			condition.wait(lock);
		}

		// pausa da 2 a 4 secondi
		uniform_int_distribution<int> dist(2, 4);
		seconds duration(dist(gen));
		sleep_for(duration);

		bales++;
		cout << tname << ": ci sono " << bales << " balle di fieno." << endl;
		condition.notify_one();
	}
}

void eat(int& bales, mutex& _mutex, condition_variable& condition, const char* tname)
{
	static thread_local default_random_engine gen(random_device{}());

	for (int i = 0; i < 3; i++) {
		unique_lock<mutex> lock(_mutex);

		while (bales == 0)
		{
			condition.wait(lock);
		}

		// pausa da 1 a 3 secondi
		uniform_int_distribution<int> dist(1, 3);
		seconds duration(dist(gen));
		sleep_for(duration);

		bales--;
		cout << tname << ": ci sono " << bales << " balle di fieno." << endl;
		condition.notify_one();
	}
}

int main()
{
	int bales = 0;
	mutex _mutex;
	condition_variable condition;

	thread farmer(heap, ref(bales), ref(_mutex), ref(condition), "contadino");
	thread cow(eat, ref(bales), ref(_mutex), ref(condition), "mucca");

	farmer.join();
	cow.join();
}

File: IT/Multithreading/Java/Farm.java
/* file Farm.java */

import java.util.concurrent.ThreadLocalRandom;

/* classe che rappresenta il fienile */

class Barn {
	private volatile int bales = 0;

	public synchronized void heap() {
		for (int i = 0; i < 3; i++) {
			try {
				while (bales > 0) {
					wait();
				}

				Thread t = Thread.currentThread();
				// pausa da 2 a 4 secondi
				int pause = ThreadLocalRandom.current().nextInt(2, 4+1);
				t.sleep(pause*1000);

				bales++;
				System.out.println(t.getName() + ": ci sono " + bales + " balle di fieno.");
				notify();
			}
			catch (InterruptedException e) {
				System.err.println(e);
			}
		}
	}

	public synchronized void eat() {
		for (int i = 0; i < 3; i++) {
			try {
				while (bales == 0) {
					wait();
				}

				Thread t = Thread.currentThread();
				// pausa da 1 a 3 secondi
				int pause = ThreadLocalRandom.current().nextInt(1, 3+1);
				t.sleep(pause*1000);

				bales--;
				System.out.println(t.getName() + ": ci sono " + bales + " balle di fieno.");
				notify();
			}
			catch (InterruptedException e) {
				System.err.println(e);
			}
		}
	}
}

/* classe che rappresenta un contadino */

class Farmer implements Runnable {
	private Barn barn;

	public Farmer(Barn barn) {
		this.barn = barn;
	}

	@Override
	public void run() {
		barn.heap();
	}
}

/* classe che rappresenta una mucca */

class Cow implements Runnable {
	private Barn barn;
	
	public Cow(Barn barn) {
		this.barn = barn;
	}

	@Override
	public void run() {
		barn.eat();
	}
}

/* la classe fattoria contiene il metodo main() */

public class Farm {
	public static void main(String [] args) {
		Barn barn = new Barn();

		Thread farmer = new Thread(new Farmer(barn), "contadino");
		Thread cow = new Thread(new Cow(barn), "mucca");

		farmer.start();
		cow.start();
	}
}

